File: 0108 - Convert Sorted Array to Binary Search Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, A: List[int]) -> Optional[TreeNode]:
        def solve(l, r):
            if l > r:
                return None
            mid = l+(r-l)//2
            return TreeNode(A[mid], solve(l, mid-1), solve(mid+1, r))
        return solve(0, len(A)-1)

File: 0108 - Convert Sorted Array to Binary Search Tree/solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* newNode(int val, struct TreeNode *left, struct TreeNode *right){
    struct TreeNode *res = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    res->val = val, res->left = left, res->right = right;
    return res;
}
struct TreeNode* solve(int *A, int l, int r){
    if(l > r) return NULL;
    int mid = l+(r-l)/2;
    return newNode(A[mid], solve(A, l, mid-1), solve(A, mid+1, r));
}
struct TreeNode* sortedArrayToBST(int* A, int ASz){
    return solve(A, 0, ASz-1);
}